### ISSUE TITLE
the mask in your implementation is inconsistent with your paper

### ISSUE THREAD

**Issue by jxs129** [dydjw9/Efficient_SAM#5]
Based on your implementation `Esam.py`,  same parameters are randomly masked at the descent step (i.e., the second step as follows).
But in your paper (the paragraph above eq(5)), the mask is added to the perturbance $\epsilon$.
As they are very different, could you please provide some explanations?

```
@torch.no_grad()
    def second_step(self, zero_grad=False):
        for group in self.param_groups:
            for p in group["params"]:
                if p.grad is None or not self.state[p]: continue
                p.sub_(self.state[p]["e_w"])  # get back to "w" from "w + e(w)"
                self.state[p]["e_w"] = 0

                if random.random() > self.beta:
                    p.requires_grad = False

        self.base_optimizer.step()  # do the actual "sharpness-aware" update

        if zero_grad: self.zero_grad()
```

**Comment by dydjw9** [comment]
Hello,  thank you for your question. Note that the p.requires_grad = False is set after the second .backward() operation. That means all of the parameters' gradients will be calculated and stored as p.grad. Therefore, the base_optimizer.step() will update all the parameters, as each parameter has grad. That is the feature of pytorch. 

The setting of p.requires_grad = False will only affect the first .backward() operation. We do it here because we do not need to traverse all the parameters one more time. 

Is my explanation clear? 

Best regards

**Comment by jxs129** [comment]
thanks for your response.
GOOD tricks in the implementation.

one follow-up question:
in the above code block, `p` is a tensor (e.g., a weight matrix with $64\times 64\times 3\times 3$ in a conv layer),
which means the whole matrix (i.e., `p`) is masked,  not some of the parameters in `p`, right?

**Comment by dydjw9** [comment]
Yes, correct. As a tensor is the smallest unit we can operate in Pytorch framework. If we only mask some parameters of p, the performance improvement is still effective, but the computational efficiency will not be improved.
